src/imperative/cached_op.cc

```diff
@@ -212,82 +223,62 @@ nnvm::Graph Imperative::CachedOp::GetForwardGraph(
-            recording ? "full_ref_count" : "forward_ref_count"));
-    g.attrs[recording ? "full_mem_plan" : "forward_mem_plan"] =
-        std::make_shared<dmlc::any>(std::move(mem_plan));
-  };
 
 #if MXNET_USE_NGRAPH == 1
-  if (!recording) {
     ngraph_bridge::BindArgBase bind(num_inputs());
     ngraph_bridge::Compiler compiler(
-        fwd_graph_, inputs[0]->ctx(), cached_shape_inputs, cached_dtype_inputs,
+        fwd_graph_orig_, inputs[0]->ctx(), inputs_, cached_shape_inputs, cached_dtype_inputs,
         cached_storage_type_inputs);
 
-    ngraph_fwd_graph_ = compiler.Compile();
-    CheckAndInferShape(&ngraph_fwd_graph_, std::move(cached_shape_inputs),
+    g = compiler.Compile();
+    CheckAndInferShape(&g, std::move(cached_shape_inputs),
                        true);
-    CheckAndInferType(&ngraph_fwd_graph_, std::move(cached_dtype_inputs), true);
+    CheckAndInferType(&g, std::move(cached_dtype_inputs), true);
     exec::DevMaskVector cached_dev_mask(
-        ngraph_fwd_graph_.indexed_graph().num_nodes(),
+        g.indexed_graph().num_nodes(),
         inputs[0]->ctx().dev_mask());
-    CheckAndInferStorageType(&ngraph_fwd_graph_, std::move(cached_dev_mask),
+    CheckAndInferStorageType(&g, std::move(cached_dev_mask),
                              std::move(cached_storage_type_inputs), true);
 
-    const auto& idx = ngraph_fwd_graph_.indexed_graph();
-
+    const auto& idx = g.indexed_graph();
     std::vector<uint32_t> ref_count(idx.num_node_entries(), 0);
     for (const auto& i : idx.input_nodes()) ++ref_count[idx.entry_id(i, 0)];
     for (const auto& i : idx.outputs()) ++ref_count[idx.entry_id(i)];
     for (size_t i = 0; i < idx.num_nodes(); ++i) {
       for (const auto& j : idx[i].inputs) ++ref_count[idx.entry_id(j)];
     }
-
-    ngraph_fwd_graph_.attrs["forward_ref_count"] =
+    g.attrs["forward_ref_count"] =
         std::make_shared<dmlc::any>(ref_count);
-    create_memory_plan(ngraph_fwd_graph_);
-    return ngraph_fwd_graph_;
+
+    full_graph_ = GetFullGraph(g, compiler.GetInputs());
+  StorageVector storage(idx.num_node_entries(), exec::kBadStorageID);
+  for (const auto i : idx.input_nodes()) storage[idx.entry_id(i, 0)] = exec::kExternalStorageID;
+  const auto& stypes = g.GetAttr<StorageTypeVector>("storage_type");
+  CHECK_EQ(stypes.size(), storage.size());
+  for (size_t i = 0; i < stypes.size(); i++) {
+    if (stypes[i] != kDefaultStorage)
+      storage[i] = exec::kDynamicStorageID;
   }
 #endif
-
-  create_memory_plan(g);
+  auto storage_copy = storage;
+  auto mem_plan = PlanMemory(
+      &g, std::move(storage), g.GetAttr<std::vector<uint32_t> >(
+          "full_ref_count"));
+  g.attrs["full_mem_plan"] =
+      std::make_shared<dmlc::any>(std::move(mem_plan));
+  mem_plan = PlanMemory(
+      &g, std::move(storage_copy), g.GetAttr<std::vector<uint32_t> >(
+         "forward_ref_count"));
+  g.attrs["forward_mem_plan"] =
+      std::make_shared<dmlc::any>(std::move(mem_plan));
```

Comment:
Why do you need to create both memory plans, cant you not use the `recording ? "full_ref_count" : "forward_ref_count` trick?

Reply:
mainly for perf.optimization, this avoid recompile.